rfcs/text/0004_application_service_mounting.md

```diff
@@ -0,0 +1,238 @@
+- Start Date: 2019-05-10
+- RFC PR: (leave this empty)
+- Kibana Issue: (leave this empty)
+
+# Summary
+
+A front-end service to manage registration and root-level routing for
+first-class applications.
+
+# Basic example
+
+```tsx
+import React from 'react';
+import ReactDOM from 'react-dom';
+
+class MyPlugin {
+  setup({ application }) {
+    application.register({
+      id: 'my-app',
+      title: 'My Application',
+      async mount(context, targetDomElement) {
+        const { MyApp } = await import('./components');
+
+        ReactDOM.render(
+          <MyApp mountContext={context} deps={pluginStart} />,
+          targetDomElement
+        );
+
+        return () => {
+          ReactDOM.unmountComponentAtNode(targetDomElement);
+        };
+      }
+    });
+  }
+}
+```
+
+# Motivation
+
+By having centralized management of applications we can have a true single page
+application. It also gives us a single place to enforce authorization and/or
+licensing constraints on application access.
+
+By making the mounting interface of the ApplicationService generic, we can
+support many different rendering technologies simultaneously to avoid framework
+lock-in.
+
+# Detailed design
+
+This is the bulk of the RFC. Explain the design in enough detail for somebody
+familiar with Kibana to understand, and for somebody familiar with the
+implementation to implement. This should get into specifics and corner-cases,
+and include examples of how the feature is used. Any new terminology should be
+defined here.
+
+## Interface
+
+```ts
+/** A context type that implements the Handler Context pattern from RFC-0003 */
+export interface MountContext {
+  core: {
+    chrome: ChromeStart;
```

Comment:
This isn't necessarily critical for this RFC, but in the spirit of the RFC demonstrating real world examples, I figured I'd mention it. I leave it to you whether you change that here or not.

I don't think the application mounting function should get access to the `start` contracts of plugins. In fact, I suspect the application mount handler would get access to _more_ things than would be exposed via plugin start. We expose basically everything we can to the plugin lifecycle hooks today, but with the introduction of the handler proposal we should be able to scale back the capabilities exposed at a top level in favor of those exposed directly to handlers.

Reply:
Agreed, many of these services do not make sense outside the scope of rendering (for example, overlays). I was expecting to move these out of CoreStart as well. I'll go ahead and update these to not reference `*Start` types since those will not be present in the future, making this RFC confusing.